linera-core/src/client/mod.rs

```diff
@@ -422,6 +423,58 @@ where
     }
 }
 
+// Tracks which chain_id<>block_height pairs have been successfully processed.
+struct MultichainTracker {
+    // Starting tracker index.
+    tracker: u64,
+    // Tracks the highest block height processed for each chain.
+    highest_processed: BTreeMap<ChainId, BlockHeight>,
+}
+
+impl MultichainTracker {
+    fn new(init: u64) -> Self {
+        Self {
+            tracker: init,
+            highest_processed: BTreeMap::new(),
+        }
+    }
+
+    fn push(&mut self, chain_id: ChainId, height: BlockHeight) {
+        match self.highest_processed.entry(chain_id) {
+            btree_map::Entry::Occupied(mut entry) => {
+                if *entry.get() < height {
+                    entry.insert(height);
+                }
+            }
+            btree_map::Entry::Vacant(vacant) => {
+                vacant.insert(height);
+            }
+        }
+    }
+
+    // Returns an index of the last processed certificate in the original remote_node_received_log.
+    //
+    // Goes through the original `remote_node_received_log` and checks which of the certificates have been received.
+    // Upon finding the first entry that has not been processed, it stops and returns the index of the last processed certificate.
```

Comment:
We could add `///` to all of these comments, in case we make any of them public. (And I think `rustdoc` even has an option to render private docs.)

Reply:
Done in [c660103](https://github.com/linera-io/linera-protocol/pull/2666/commits/c66010359a10c4462def29e143edf267aec539dc).